route-executor/src/lib.rs

```diff
@@ -0,0 +1,269 @@
+        UnexpectedError,
+    }
+
+    #[pallet::call]
+    impl<T: Config> Pallet<T> {
+        /// Executes a sell with a series of trades specified in the route.
+        /// The price for each trade is determined by the corresponding AMM.
+        ///
+        /// - `origin`: The executor of the trade
+        /// - `asset_in`: The identifier of the asset to sell
+        /// - `asset_out`: The identifier of the asset to receive
+        /// - `amount_in`: The amount of `asset_in` to sell
+        /// - `limit`: The minimum amount of `asset_out` to receive.
+        /// - `route`: Series of trades containing AMM and asset pair information
+        ///
+        /// Emits `RouteIsExecuted` when successful.
+        #[pallet::weight(<T as Config>::WeightInfo::execute_sell(route.len() as u32))]
+        pub fn execute_sell(
+            origin: OriginFor<T>,
+            asset_in: T::AssetId,
+            asset_out: T::AssetId,
+            amount_in: T::Balance,
+            limit: T::Balance,
+            route: Vec<Trade<T::AssetId>>,
+        ) -> DispatchResult {
+            let who = ensure_signed(origin)?;
+            Self::validate_route_size(route.len())?;
+
+            ensure!(
+                T::Currency::reducible_balance(asset_in, &who, false) >= amount_in,
+                Error::<T>::InsufficientBalance
+            );
+
+            let mut amounts_to_sell = Vec::<AmountWithFee<T::Balance>>::with_capacity(route.len() + 1);
+            let mut amount = AmountWithFee::new_without_fee(amount_in);
+            amounts_to_sell.push(amount);
+
+            for trade in route.iter() {
+                let result = T::AMM::calculate_sell(trade.pool, trade.asset_in, trade.asset_out, amount);
+                match result {
+                    Err(ExecutorError::NotSupported) => return Err(Error::<T>::PoolNotSupported.into()),
+                    Err(ExecutorError::Error(dispatch_error)) => return Err(dispatch_error),
+                    Ok(amount_to_sell) => {
+                        amount = amount_to_sell;
+                        amounts_to_sell.push(amount_to_sell);
+                    }
+                }
+            }
+
+            let last_amount = amounts_to_sell.pop().ok_or(Error::<T>::UnexpectedError)?;
+            ensure!(last_amount.amount >= limit, Error::<T>::MinLimitToReceiveNotReached);
+
+            for (amount, trade) in amounts_to_sell.iter().zip(route) {
+                let execution_result = T::AMM::execute_sell(trade.pool, &who, trade.asset_in, trade.asset_out, *amount);
+
+                match execution_result {
+                    Ok(_) => continue,
+                    Err(err) => {
+                        return match err {
+                            ExecutorError::NotSupported => Err(Error::<T>::PoolNotSupported.into()),
+                            ExecutorError::Error(dispatch_error) => Err(dispatch_error),
+                        }
+                    }
+                }
```

Comment:
no need for continue statement.

```suggestion
                 if let Err(error) = execution_result {
                    return match error {
                            ExecutorError::NotSupported => Err(Error::<T>::PoolNotSupported.into()),
                            ExecutorError::Error(dispatch_error) => Err(dispatch_error),
                        }
                }
```

Reply:
actually you could even combine and map err in the execute call.

Reply:
The first option looks fine, good one!

But unfortunately I could not combine it to
```

                T::AMM::execute_buy(trade.pool, &who, trade.asset_in, trade.asset_out, *amount)
                    .map_err(|err| match err {
                        ExecutorError::NotSupported => Err(Error::<T>::PoolNotSupported.into()),
                        ExecutorError::Error(dispatch_error) => Err(dispatch_error),
                    })?;
```

as it complains with error
```
error[E0277]: `?` couldn't convert the error to `sp_runtime::DispatchError`
   --> route-executor/src/lib.rs:237:23
    |
237 |                     })?;
    |                       ^ the trait `From<Result<_, sp_runtime::DispatchError>>` is not implemented for `sp_runtime::DispatchError`
```